### ISSUE TITLE
Use a newer version of LLDB for debugging?

### ISSUE THREAD

**Issue by iKlask** [bytecodealliance/wasm-micro-runtime#3622]
So it seems that debugging wasm code in WAMR relies on an older version of LLDB (v13.x) with a [specific patch applied](https://reviews.llvm.org/D78801).

Having to pull an older version of LLVM, apply this patch, and build it yourself is quite a hassle, especially since WAMR doesn't supply prebuilt binaries for Windows (only for Linux and macOS). Moreover, it seems unlikely that this patch will ever be integrated into the main LLVM branch.

So my question: Have modern versions of LLVM/LLDB supplied features this patch attempted to add, or is this patch is still necessary today for debugging WASM in WAMR? I am asking because I've noticed that the Wasmtime runtime works well with more modern builds of LLDB without requiring any patches.

**Comment by yamt** [comment]
> So my question: Have modern versions of LLVM/LLDB supplied features this patch attempted to add, or is this patch still necessary today for debugging WASM in WAMR?

the latter.

>  I am asking because I've noticed that the Wasmtime runtime works well with more modern builds of LLDB without requiring any patches.

wamr aot-debug doesn't require the patch. (it has its own problems though)

**Comment by srberard** [comment]
I'd like to revisit this topic.  As pointed out above, we still need the patch and it's based on a version of LLVM that is several years old.  It would be ideal to not have to have a patch to support debugging interpreted Wasm code.  Also mentioned above, this works out of the box with WasmTime today with no need for a patched version of LLDB.  I'm not sure how this is done in WasmTime, but it proves that this can be done and the experience is excellent.

My desire here is to improve the developer experience and have a good debugging experience.  While I am able to get the existing debugger built and working, I've hit a number of issues and bugs.  Fixing them on the old codebase doesn't seem to be the right thing to do.  

From what I understand, a key issue was with the fact that Wasm supports multiple memories and that this was not handled well in LLDB at the time.  Looking through the patch in the WAMR codebase, if I'm understanding this correctly, a plugin was created for handling this but that required a number of changes in core parts of LLDB.  In the end, it was never upstreamed into LLVM.  Now this was all done a while ago (LLVM 13.x if memory serves) and a lot has changed since then.  I've seen several active issues around Wasm debugging support in the LLVM project including [this](https://github.com/llvm/llvm-project/issues/150449) issue and [this](https://github.com/llvm/llvm-project/issues/151439) issue.  There's clearly interest in making this work.

Would it be possible to revisit this and see what can be done to align things so that debugging works out of the box with WAMR and LLDB?

**Comment by JDevlieghere** [comment]
ToT LLDB supports debugging Wasm running under `iwasm` without any patches. I gave a talk about this at the [LLVM Devroom at FOSDEM](https://fosdem.org/2026/schedule/event/87ZLQV-wasm-debugging-lldb/). We should be able to remove the old instructions once LLVM 22 is released.

**Comment by iKlask** [comment]
> ToT LLDB supports debugging Wasm running under `iwasm` without any patches. I gave a talk about this at the [LLVM Devroom at FOSDEM](https://fosdem.org/2026/schedule/event/87ZLQV-wasm-debugging-lldb/). We should be able to remove the old instructions once LLVM 22 is released.

Oh wow cool! Good talk! I have an embedded WASM project with WAMR i've put off from dealing with for 2 years because I didnt enjoy the debug experience. This may get me back into the project haha. I see 22.1.0 is out now and in the release notes mentions:

> - LLDB can now set breakpoints, show backtraces, and display variables when debugging Wasm with supported runtimes (WAMR and V8).
> - LLDB now has a Wasm platform, which can be configured to run WebAssembly binaries directly under a Wasm runtime. Configurable through the platform.plugin.wasm settings.

So with the latest LLDB I should be able to connect to the iwasm debug stub?

**Comment by JDevlieghere** [comment]
> So with the latest LLDB I should be able to connect to the iwasm debug stub?

Yes, that's right. Everything that used to work with the patched version of LLDB now works out of the box. With the notable exception of expression evaluation, I've manually verified most basic debugging scenarios. The next step for us, as I mentioned in the talk, will be setting up automated testing so we can get coverage for both LLDB and the GDB stubs.
